platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleTreeSnapshotReader.java

```diff
@@ -0,0 +1,166 @@
+                    final int fileVersion = readAndCheckStateFileVersion(in);
+
+                    final Path directory = stateFile.getParent();
+                    if (fileVersion == INIT_STATE_FILE_VERSION) {
+                        return readStateFileDataV1(stateFile, in, directory);
+                    } else if (fileVersion == SIG_SET_SEPARATE_STATE_FILE_VERSION) {
+                        return readStateFileDataV2(stateFile, in, directory);
+                    } else {
+                        throw new IOException("Unsupported state file version: " + fileVersion);
+                    }
+                });
+    }
+
+    /**
+     * This method reads the state file data from a version 1 state file. This version of the state file contains
+     * signature set data.
+     */
+    @NonNull
+    private static StateFileData readStateFileDataV1(
+            @NonNull final Path stateFile, @NonNull final MerkleDataInputStream in, @NonNull final Path directory)
+            throws IOException {
+        try {
+            final MerkleRoot state = in.readMerkleTree(directory, SignedStateFileUtils.MAX_MERKLE_NODES_IN_STATE);
+            final Hash hash = in.readSerializable();
+            final SigSet sigSet = in.readSerializable();
+            return new StateFileData(state, hash, sigSet);
+        } catch (final IOException e) {
+            throw new IOException("Failed to read snapshot file " + stateFile.toFile(), e);
+        }
+    }
+
+    /**
+     * This method reads the state file data from a version 2 state file. This version of the state file
+     * doesn't contain signature set data. Instead, the signature set data is stored in a separate file,
+     * and the resulting object doesn't have {@link SigSet} field initialized.
+     */
+    @NonNull
+    private static StateFileData readStateFileDataV2(
+            @NonNull final Path stateFile, @NonNull final MerkleDataInputStream in, @NonNull final Path directory)
+            throws IOException {
+        try {
+            final MerkleRoot state = in.readMerkleTree(directory, SignedStateFileUtils.MAX_MERKLE_NODES_IN_STATE);
+            final Hash hash = in.readSerializable();
+            return new StateFileData(state, hash, null);
+
+        } catch (final IOException e) {
+            throw new IOException("Failed to read snapshot file " + stateFile.toFile(), e);
+        }
+    }
+
+    /**
+     * Read the version from a signed state file and check it
+     *
+     * @param in the stream to read from
+     * @throws IOException if the version is invalid
+     * @return the protocol version
+     */
+    private static int readAndCheckStateFileVersion(@NonNull final MerkleDataInputStream in) throws IOException {
+        final byte versionByte = in.readByte();
+        if (versionByte != VERSIONED_FILE_BYTE) {
+            throw new IOException("File is not versioned -- data corrupted or is an unsupported legacy state");
```

Comment:
The javadoc for `VERSIONED_FILE_BYTE` says `The signed state file was not versioned before, this byte was introduced to mark a versioned file`.

I'm wondering if the `before` here refers to the v1 or something prior to the v1? If it does refer to the v1, then I don't think we should throw an exception here. If it refers to something before the v1 even, then I suggest to articulate this more clearly in the javadoc for this constant.

Reply:
>I'm wondering if the before here refers to the v1 or something prior to the v1? 

I think it was prior to v1. This particular byte has been there for many releases. 

>then I suggest to articulate this more clearly in the javadoc for this constant.

how would you recommend to rephrase it?

Reply:
Something like:

```
Prior to v1, the signed state file was not versioned. This byte was introduced in v1 to mark a versioned file.
```

Reply:
Thank you. I updated javadoc as you suggested.